[Core] Eliminate redundant is_encoder_decoder lookups (>20us/token)

wushidonguc opened the pull request:
Summary:
Cache `model_config.is_encoder_decoder` once at the start of `_preprocess()` instead of accessing it multiple times within the method to eliminate redundant property lookups.

**Impact:** Saves >20 microseconds per inference step by eliminating redundant property lookups in the hot path.

**Performance Evidence:**
- Before: [Trace showing baseline performance](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117471%2Flocalhost%2Flibkineto_activities_4057258.json.gz#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117471%2Flocalhost%2Flibkineto_activities_4057258.json.gz&local_cache_key)
 {F1983818516}
<img width="2328" height="1511" alt="is_encoder_decoder before" src="https://github.com/user-attachments/assets/852073b7-8be8-4b99-aeac-b76870f84640" />

- After: [Trace showing optimized performance](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117915%2Flocalhost%2Flibkineto_activities_364733.json.gz#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117915%2Flocalhost%2Flibkineto_activities_364733.json.gz&local_cache_key)
 {F1983814990}
<img width="2329" height="1514" alt="is_encoder_decoder after" src="https://github.com/user-attachments/assets/2895ca5e-d703-4ec3-a6be-ccafe38be95d" />


Test Plan: CI

Differential Revision: D87902650

njhill reviewed (commented):
nit: both of these if's could fit on a single line now...
```python
        if self.supports_mm_inputs and is_first_rank and not is_encoder_decoder:
            # Run the multimodal encoder if any.
            with self.maybe_get_ec_connector_output(
```

wushidonguc committed:
Eliminate redundant is_encoder_decoder lookups (20-40us/step) (#29800)

Summary:

Cache `model_config.is_encoder_decoder` once at the start of `_preprocess()` instead of accessing it multiple times within the method to eliminate redundant property lookups.

**Impact:** Saves 20-40 microseconds per inference step by eliminating redundant property lookups in the hot path.

**Performance Evidence:** 
- Before: [Trace showing baseline performance](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117471%2Flocalhost%2Flibkineto_activities_4057258.json.gz#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117471%2Flocalhost%2Flibkineto_activities_4057258.json.gz&local_cache_key)
 {F1983818516} 
- After: [Trace showing optimized performance](https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117915%2Flocalhost%2Flibkineto_activities_364733.json.gz#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Fgpu_traces%2Ftree%2Ftraces%2Fdynocli%2F0%2F1764117915%2Flocalhost%2Flibkineto_activities_364733.json.gz&local_cache_key) 
 {F1983814990}

Signed-off-by: Wushi Dong <dongws@meta.com>

Test Plan: CI

Reviewed By: helunwencser

Differential Revision: D87902650

wushidonguc commented:
Fix pre-commit check

DarkLight1337 committed:
Merge branch 'main' into export-D87902650
